google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/main/java/com/google/cloud/spanner/jdbc/JdbcTypeConverter.java

```diff
@@ -0,0 +1,274 @@
+        if (type.getCode() == Code.INT64) return BigDecimal.valueOf((Long) value);
+        if (type.getCode() == Code.FLOAT64) return BigDecimal.valueOf((Double) value);
+        if (type.getCode() == Code.STRING) return new BigDecimal((String) value);
+      }
+      if (targetType.equals(Long.class)) {
+        if (type.getCode() == Code.BOOL) return (Boolean) value ? 1L : 0L;
+        if (type.getCode() == Code.INT64) return value;
+        if (type.getCode() == Code.FLOAT64) return ((Double) value).longValue();
+        if (type.getCode() == Code.STRING) return Long.valueOf((String) value);
+      }
+      if (targetType.equals(Integer.class)) {
+        if (type.getCode() == Code.BOOL) return (Boolean) value ? 1 : 0;
+        if (type.getCode() == Code.INT64) return ((Long) value).intValue();
+        if (type.getCode() == Code.FLOAT64) return ((Double) value).intValue();
+        if (type.getCode() == Code.STRING) return Integer.valueOf((String) value);
+      }
+      if (targetType.equals(BigInteger.class)) {
+        if (type.getCode() == Code.BOOL) return (Boolean) value ? BigInteger.ONE : BigInteger.ZERO;
+        if (type.getCode() == Code.INT64) return BigInteger.valueOf((Long) value);
+        if (type.getCode() == Code.FLOAT64) return BigInteger.valueOf(((Double) value).longValue());
+        if (type.getCode() == Code.STRING) return new BigInteger((String) value);
+      }
+      if (targetType.equals(Float.class)) {
+        if (type.getCode() == Code.BOOL)
+          return (Boolean) value ? Float.valueOf(1f) : Float.valueOf(0f);
+        if (type.getCode() == Code.INT64) return ((Long) value).floatValue();
+        if (type.getCode() == Code.FLOAT64) return ((Double) value).floatValue();
+        if (type.getCode() == Code.STRING) return Float.valueOf((String) value);
+      }
+      if (targetType.equals(Double.class)) {
+        if (type.getCode() == Code.BOOL)
+          return (Boolean) value ? Double.valueOf(1d) : Double.valueOf(0d);
+        if (type.getCode() == Code.INT64) return ((Long) value).doubleValue();
+        if (type.getCode() == Code.FLOAT64) return value;
+        if (type.getCode() == Code.STRING) return Double.valueOf((String) value);
+      }
+    } catch (Exception e) {
+      throw JdbcSqlExceptionFactory.of(
+          "Cannot convert " + value + " to " + targetType.getName(),
+          com.google.rpc.Code.INVALID_ARGUMENT,
+          e);
+    }
+
+    throw JdbcSqlExceptionFactory.of(
+        "Cannot convert " + type.getCode().name() + " to " + targetType.getName(),
+        com.google.rpc.Code.INVALID_ARGUMENT);
+  }
+
+  @SuppressWarnings("deprecation")
+  static Date toGoogleDate(java.sql.Date date) {
+    return date == null
+        ? null
+        : Date.fromYearMonthDay(date.getYear() + 1900, date.getMonth() + 1, date.getDate());
+  }
+
+  static Date toGoogleDate(java.sql.Time date) {
+    return Date.fromYearMonthDay(1970, 1, 1);
+  }
+
+  @SuppressWarnings("deprecation")
```

Comment:
what's deprecated here?

Reply:
Specifically in this place: `java.sql.Date#getYear()`, `java.sql.Date#getMonth()` and `java.sql.Date#getDate()`.

More in general: Many methods in `java.sql.Date` and `java.sql.Timestamp` have been deprecated since JDK version 1.1. They made some very bad choices regarding timezone handling in these classes, but they are still used as both return types and parameter types in the JDBC interfaces.

Cloud Spanner does not support storing timestamps with timezone information in the database, so the mismatch with these classes is a little bit less of a problem for Cloud Spanner than for many other DBMS'es.